### ISSUE TITLE
Code Action provider called n times on a single key stroke

### ISSUE THREAD

**Issue by dbaeumer** [microsoft/vscode#74952]
I noticed today that my ESLint code action provider is called n times on every keystroke. This only seems to happen in a build version of VS Code. I couldn't reproduce this out of source. So I suspect a timing / debouncing issues. I debugged it in the built version and here is what is happening:

- in the test workspace this is happening the following LS provide diagnostics: tslint, TS and ESLint
- on a keystroke all three LS update the diagnostics
- this results to 3 calls to `_onMarkerChanges` here: https://github.com/Microsoft/vscode/blob/master/src/vs/editor/contrib/codeAction/codeActionModel.ts#L45
- depending on the timeing these seem to get folded or not

The more general question for me is why the ESLint code action provider is called when TS or TSLint diagnostics change. I do already check if the diagnostics contain ESLint errors. However in the above example if the ESLint errors come in first then there are errors when the code action provider is triggered again on a TSLint or TS diagnostic change.

**Comment by mjbvz** [comment]
We currently have no way to associate diagnostics with a code action provider. That is all left up to the provider itself

 It sounds like what you'd like is a way for a code action provider to say: I'm only interested in diagnostics of `"source": "eslint"`, or something like that? The provider would still have to handle being triggered with unexpected diagnostics but that way we may be able to avoid some extra calls

**Comment by dbaeumer** [comment]
Yes, what I would like to code is the following: only call the code action provider is a diagnostic of type `eslint` as changed. I am pretty sure that none of the code action provider has a check to ensure it is not computing the actions again if none of it diagnostic have changed since the last call.

**Comment by dbaeumer** [comment]
@mjbvz can you please explain why you think this shouldn't be handled VS Code core. The is a unnecessary re-computation of  quick fixes which IMO the core should handle. Asking every provider to handle this will not happen since most providers are not even aware of this.

Can you please consider to reopen this or at least explain why this can't be addressed in VS Code core.

**Comment by mjbvz** [comment]
I did not see us getting to this soon but we can keep tracking it

**Comment by dbaeumer** [comment]
Thanks!
